### PR TITLE
[MultiGPU] Support pre-sharded model weights

### PR DESCRIPTION
**Pull request by Lunderberg**
Prior to this PR, sharding of model weights was done as the model is being initialized.  This allows a single set of model weights to be used regardless of the number of GPUs available, but may have slow model start-up time, due to the processing performed during start-up.  This processing is done by GPU-0, which then scatters the resulting model weights to all other GPUs.

This PR allows generation and use of pre-sharded model weights.  Here, the sharding of weight tensors is performed at model-build time, removing the run-time computations.  In addition, because each sharded weight tensor is only required by a specific worker, each worker can directly load their subset of the model weights.

The improvement in startup time when compared to initialization-time sharding depends on the number of GPUs being used.  This is due both to the decreased initialization time with pre-sharded weights (smaller sharded size per GPU), and the increased initialization time with un-sharded weights (GPU-0 must scatter data to more GPUs).

### CODE REVIEW

**Comment by Lunderberg** [comment]
This change requires support in TVM Unity.  [TVM#15916](https://github.com/apache/tvm/pull/15916), [TVM#15942](https://github.com/apache/tvm/pull/15942), and [TVM#15957](https://github.com/apache/tvm/pull/15957) are required to build/execute the model.  [TVM#15941](https://github.com/apache/tvm/pull/15941) is not strictly required, but improves the GPU memory footprint while converting the weights.  This change should not be merged into MLC-LLM until the upstream prerequisites have been landed.

In the meantime, the TVM dev branch [here](https://github.com/Lunderberg/tvm/tree/dev_pre_sharded_weights) contains the four upstream PRs, and can be used to test the MLC-LLM changes.

**Comment by masahi** [comment]
@Lunderberg Can you rebase? I want to merge this into our branch to work on a feature that depends on build-time sharding.

**Comment by Lunderberg** [comment]
The PR is now rebased, and has been verified to work locally.

* [TVM#15942](https://github.com/apache/tvm/pull/15942): landed
* [TVM#15957](https://github.com/apache/tvm/pull/15957): waiting on review
* [TVM#15941](https://github.com/apache/tvm/pull/15941): landed

* [TVM#15916](https://github.com/apache/tvm/pull/15916): No longer a blocking requirement.  The underlying bug still exists (downstream Relax transforms relying on unchecked assumptions) is still present, but is no longer being triggered when building the model weights.  (It will also likely be replaced with a different long-term fix, see link for details.)

File: mlc_llm/core.py
Comment (masahi) [inline mlc_llm/core.py:284]: Better to call it `use_presharded_weight` to align with `convert_weight_only`

File: mlc_llm/utils.py
Comment (Lunderberg) [reply mlc_llm/utils.py:283]: Thank you, and removed.

File: mlc_llm/utils.py
Comment (Lunderberg) [reply mlc_llm/utils.py:287]: Thank you, and removed.

File: mlc_llm/core.py
Comment (Lunderberg) [reply mlc_llm/core.py:831]: Thank you, and removed.

File: mlc_llm/core.py
Comment (Lunderberg) [reply mlc_llm/core.py:385]: Thank you, and updated.  Since `mlc-llm` currently doesn't have a CI lint step, I've been disabling my local use of `black` to avoid causing lots of re-formating churn.  Unfortunately, that means I've also been contributing to the problem of having unformatted code, and I don't yet have a solution to only format new code.

File: mlc_llm/core.py
Comment (Lunderberg) [reply mlc_llm/core.py:284]: Hmm.  The current use of `weight` has been one of my most frequent typos, and I had hoped to avoid adding to that problem.  Since most models have more than a single parameter (citation needed), having the singular "weight" instead of "weights" always gives me a double-take.

What if we keep the plural `--use-presharded-weights`, and in a follow-up PR add an option for `--convert-weights-only` (plural)?  To maintain backwards compatibility, both `--convert-weights-only` and `--convert-weight-only` would be accepted by the argument parser, and would share the same the same argparse destination variable to be treated the same by the rest of the script.  That would allow consistency with the other arguments (presharded weights as a plural noun), while maintaining backwards compatibility.

Comment (masahi) [reply mlc_llm/core.py:284]: Sounds good, I also prefer a plural. If you can update to --convert-weights-only while keeping backwad compat, that would be the best!

File: mlc_llm/core.py
Comment (masahi) [inline mlc_llm/core.py:824]: require

File: mlc_llm/relax_model/commons.py
Comment (masahi) [inline mlc_llm/relax_model/commons.py:146]: Why does `build_model_only` mean `param_shape_is_already_sharded`?

File: mlc_llm/core.py
Comment (Lunderberg) [reply mlc_llm/core.py]: Thank you, and fixed.

File: mlc_llm/relax_model/commons.py
Comment (Lunderberg) [reply mlc_llm/relax_model/commons.py:146]: Currently, the initial model definition produces sharded shapes for `--build-model-only`, and unsharded shapes for `--convert-weight-only`.  (Based on [this conditional](https://github.com/mlc-ai/mlc-llm/blob/main/mlc_llm/relax_model/llama.py#L63).)  Ideally, this would be part of a later transformation applied to the the model definition, rather than part of the original model definition, and that's something I want to improve in the future.

For now, since `_get_shard_strategies` may be called either with `--build-model-only` (un-sharded weights, produce sharding functions in order to save alongside model) or with `--convert-weight-only` (pre-sharded weights, produce sharding functions in order to apply at compile time), it needs to know whether the parameter shapes in the model have been sharded or not.

Comment (masahi) [reply mlc_llm/relax_model/commons.py:146]: Thanks it makes sense, I didn't know that `num-shards` is ignored in the model def when `--convert-weight-only`.

**Comment by Lunderberg** [comment]
Rebased onto main, as local testing required compatibility fixes from #1210.  Using this PR branch and the current upstream `apache/tvm:unity` branch for local testing, both compile-time sharding and run-time sharding work well, and this PR is ready to merge.

**Review by masahi — APPROVED** [review]
I confirmed that it works. @Lunderberg please make sure to follow up with `--convert-weights-only`.

**Comment by Lunderberg** [comment]
Thank you for the reminder, and I made https://github.com/mlc-ai/mlc-llm/pull/1227 to fix the `--convert-weights-only`.

**Comment by Cydia2018** [comment]
@Lunderberg When saving model weights using `--use-presharded-weights`, the following error occurred:
`
Get old param:   0%|                                                                                                                                                                           | 0/17 [00:00<?, ?tensors/s[15:11:09] 
/xxx/tvm-unity/src/tir/transforms/storage_rewrite.cc:689: Warning: The allocation requires : 141557760 * 16 bits, which is greater than the maximum of int32. The size is cast to int64.                                                                                                                                                  

Start computing and quantizing weights... This may take a while.                                                                                                                                    
Get old param:  24%|██████████████████████████████████████▎                                                                                                                            | 4/17 [00:04<00:14,  1.10s/tensors]Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main                                                                                                                              
    return _run_code(code, main_globals, None,                                                    
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code                                                                                                                                         
    exec(code, run_globals)                                                                       
  File "/xxx/mlc-llm/mlc_llm/build.py", line 47, in <module>                                                                                                           
    main()                                                                                        
  File "/xxx/mlc-llm/mlc_llm/build.py", line 43, in main                                                                                                               
    core.build_model_from_args(parsed_args)                                                       
  File "/xxx/mlc-llm/mlc_llm/core.py", line 796, in build_model_from_args
    params = utils.convert_weights(mod_transform, param_manager, params, args)                                                                                                                      
  File "/xxx/mlc-llm/mlc_llm/utils.py", line 272, in convert_weights                                                                                                   
    vm["transform_params"]()                                                                      
  File "/xxx/tvm-unity/python/tvm/_ffi/_ctypes/packed_func.py", line 239, in __call__
    raise_last_ffi_error()                                                                        
  File "/xxx/tvm-unity/python/tvm/_ffi/base.py", line 481, in raise_last_ffi_error
    raise py_err                                                                                  
  File "/xxx/tvm-unity/src/runtime/relax_vm/vm.cc", line 634, in tvm::runtime::relax_vm::VirtualMachineImpl::InvokeClosurePacked(tvm::runtime::ObjectRef const&, tvm::r
untime::TVMArgs, tvm::runtime::TVMRetValue*)
    clo->impl.CallPacked(TVMArgs(values.data(), tcodes.data(), args.size() + 1), rv);                                                                                                               
  File "/xxx/tvm-unity/src/runtime/relax_vm/vm.cc", line 708, in tvm::runtime::relax_vm::VirtualMachineImpl::GetClosureInternal(tvm::runtime::String const&, bool)::{la
mbda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}::operator()(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*) const
    *rv = static_cast<VirtualMachineImpl*>(ctx_ptr)->InvokeBytecode(gf_idx, inputs);                                                                                                                
  File "/xxx/tvm-unity/src/runtime/relax_vm/vm.cc", line 765, in tvm::runtime::relax_vm::VirtualMachineImpl::InvokeBytecode(long, std::vector<tvm::runtime::TVMRetValue
, std::allocator<tvm::runtime::TVMRetValue> > const&)
    RunLoop();
File "/xxx/tvm-unity/src/runtime/relax_vm/vm.cc", line 890, in tvm::runtime::relax_vm::VirtualMachineImpl::RunLoop()
    this->RunInstrCall(curr_frame, instr);
  File "/xxx/tvm-unity/src/runtime/relax_vm/vm.cc", line 843, in tvm::runtime::relax_vm::VirtualMachineImpl::RunInstrCall(tvm::runtime::relax_vm::VMFrame*, tvm::runtime::relax_vm::Instruction)
    this->InvokeClosurePacked(func_pool_[instr.func_idx], args, &ret);
  File "/xxx/tvm-unity/src/runtime/library_module.cc", line 78, in tvm::runtime::WrapPackedFunc(int (*)(TVMValue*, int*, int, TVMValue*, int*, void*), tvm::runtime::ObjectPtr<tvm::runtime::Object> const&)::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}::operator()(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*) const
    LOG(FATAL) << TVMGetLastError();
tvm._ffi.base.TVMError: Traceback (most recent call last):
  5: tvm::runtime::relax_vm::VirtualMachineImpl::InvokeClosurePacked(tvm::runtime::ObjectRef const&, tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)
        at /xxx/tvm-unity/src/runtime/relax_vm/vm.cc:634
  4: tvm::runtime::relax_vm::VirtualMachineImpl::GetClosureInternal(tvm::runtime::String const&, bool)::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}::operator()(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*) const
        at /xxx/tvm-unity/src/runtime/relax_vm/vm.cc:708
  3: tvm::runtime::relax_vm::VirtualMachineImpl::InvokeBytecode(long, std::vector<tvm::runtime::TVMRetValue, std::allocator<tvm::runtime::TVMRetValue> > const&)
        at /xxx/tvm-unity/src/runtime/relax_vm/vm.cc:765
  2: tvm::runtime::relax_vm::VirtualMachineImpl::RunLoop()
        at /xxx/tvm-unity/src/runtime/relax_vm/vm.cc:890
  1: tvm::runtime::relax_vm::VirtualMachineImpl::RunInstrCall(tvm::runtime::relax_vm::VMFrame*, tvm::runtime::relax_vm::Instruction)
        at /xxx/tvm-unity/src/runtime/relax_vm/vm.cc:843
  0: tvm::runtime::WrapPackedFunc(int (*)(TVMValue*, int*, int, TVMValue*, int*, void*), tvm::runtime::ObjectPtr<tvm::runtime::Object> const&)::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}::operator()(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*) const
        at /xxx/tvm-unity/src/runtime/library_module.cc:78
  File "/xxx/tvm-unity/src/runtime/library_module.cc", line 78
TVMError: Assert fail: var_placeholder_code == 3 or var_placeholder_code == 13 or var_placeholder_code == 7 or var_placeholder_code == 4, model_layers_0_self_attn_query_key_value_proj_weight_sharding_func: Expect arg[0] to be pointer                                                   
`
Can you help me? really thanks!

**Pull request merged**
